docs/architecture.md

```diff
@@ -0,0 +1,438 @@
+# Torch-MLIR Architecture
+
+## Introduction
+
+The Torch-MLIR project provides core infrastructure for bridging between the
+Torch ecosystem and the MLIR ecosystem. Torch-MLIR does not attempt to provide a
+production end-to-end flow for PyTorch programs by itself, but is a useful
+component for constructing one.
+
+## Overview
+
+Torch-MLIR has two parts, which we call the "frontend" and "backend". These two
+halves interface at an abstraction layer which we call the "backend contract"
+and which is a normalized subset of the `torch` dialect.
+
+The frontend of Torch-MLIR is concerned with interfacing to PyTorch itself, and
+then normalizing the program to the "backend contract". This part involves build
+system complexity and exposure to PyTorch API's to get the program into the MLIR
+`torch` dialect in the first place. When we interface with TorchScript, we
+additionally have a large amount of lowering and simplification to do within
+MLIR on the `torch` dialect.
```

Comment:
Is my understanding right that the frontend will change more frequently (to track changes in PyTorch ecosystem such as TorchDynamo), however the backend contract (middle-end?) would isolate / shield the lowerings to the respective backends from being impacted too much or at all? If that is correct, maybe include a statement to clarify this?

Reply:
At this point I'm not really able to make a statement about the stability of the backend contract. In some ways it can be less stable than the frontends, at least today. For example, if PyTorch changes an op or starts using a new op in place of an old op, that will, at least today, end up immediately reflected in our backend contract. The actual code for interfacing to PyTorch doesn't need to change for that to happen.